### PR TITLE
Better-ish image handling

### PR DESCRIPTION
**Pull request by tht7**
# Checklist
- [x] I have read [CONTRIBUTING.md](./CONTRIBUTING.md)
- [x] I have described what this PR contains
- [x] This PR addresses one or more open issues that were assigned to me:
      - *list issue(s) here*
- [x] If this PR alters the UI, I have attached pictures/videos

# Pull Request Information

## About this Pull Request
This was a mess to handle but this should fix the NSFW blur and apply it on EVERY IMAGE in the app, when needed

created new CachedImageWithNsfwFilter that should be uniform in our app, it takes a url and uses our cache to load it, and blurs it according to the users wishes if you marked it as NSFW

introduced a new CachedImageProvider that can give you the uniform cached image view that's loaded from a URL and respects the nsfw setting 

Had to break the Markdown view since it wasn't flexible enough, so now the MarkdownView handles images before passing on the torch of actual markdown to the library's Markdown View

Related Issues:
#308 
#291 
#262 
#266 

## Screenshots and Videos
I won't 🤣

### CODE REVIEW

File: Mlem/Views/Shared/Markdown View.swift [hunk 1]
```diff
@@ -236,9 +238,53 @@ struct MarkdownView: View
...
+            ForEach(0...max(blocks.count, images.count), id: \.hashValue) { i in
+                if blocks.count > i {
+                    getMarkdown(text: String(blocks[i]))
+                }
+                if images.count > i {
+                    imageProvider.makeImage(url: URL(string: String(images[i].1)))
+                }
+            }
+        }
+//        for i in 0...max(blocks.count, images.count) {
```
Comment (WestonHanners) [inline Mlem/Views/Shared/Markdown View.swift:277]: is this commented code needed?

Comment (tht7) [reply Mlem/Views/Shared/Markdown View.swift]: Nope! removed!

**Comment by mormaer** [comment]
Hey @tht7 - this is nice 👍 

I don't have time to review tonight but will take a look tomorrow.

My first thoughts are that I'm wondering if we could clean it up slightly so that we always use the `CachedImageProvider`, given it can produce a view directly... it could give us a single way to treat images throughout the project if they went through a single entry point (and we could hide different behaviours behind it, not just NSFW filtering).

If we worked toward that then we could write a custom linting rule too to make sure everywhere uses it 👍

File: Mlem/Views/Shared/Markdown View.swift [hunk 2]
```diff
@@ -201,8 +202,42 @@ private extension Color {
...
+            "!["
+            Capture {
+                ZeroOrMore(.any, .reluctant)
+            }
+            "]("
+            Capture {
+                ZeroOrMore(.any, .reluctant)
+            }
+            ")"
+        }
```
Comment (EricBAndrews) [inline Mlem/Views/Shared/Markdown View.swift:223]: Nit, but for readability even simple regexes should have an explanatory comment

File: Mlem/Views/Shared/MarkdownImageProvider.swift [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        self.url = url
+        self._showNsfwFilterToggle = .init(initialValue: true)
+    }
+
+    var body: some View {
+        ZStack {
+            CachedAsyncImage(url: url, urlCache: AppConstants.urlCache) { image in
+                image
+                    .resizable()
+//                    .frame(maxWidth: .infinity)
```
Comment (EricBAndrews) [inline Mlem/Views/Shared/MarkdownImageProvider.swift:46]: Can be removed?

File: Mlem/Views/Tabs/Posts/Components/Comment/Components/Comment Item Old.swift
Comment (EricBAndrews) [inline Mlem/Views/Tabs/Posts/Components/Comment/Components/Comment Item Old.swift:1]: This whole file should be removed, I think it snuck in with one of my PRs

**Review by EricBAndrews — CHANGES_REQUESTED** [review]

**Review by WestonHanners — APPROVED** [review]

File: Mlem/Views/Shared/Markdown View.swift (hunk 2 above)
Comment (tht7) [reply Mlem/Views/Shared/Markdown View.swift]: Added comments!

File: Mlem/Views/Shared/MarkdownImageProvider.swift (hunk 3 above)
Comment (tht7) [reply Mlem/Views/Shared/MarkdownImageProvider.swift]: removed!

File: Mlem/Views/Tabs/Posts/Components/Comment/Components/Comment Item Old.swift
Comment (tht7) [reply Mlem/Views/Tabs/Posts/Components/Comment/Components/Comment Item Old.swift:1]: ahaha it was there when I got it I swear! but removed as well 😄

**Comment by tht7** [comment]
@mormaer actually, I think the opposite

 I created this image provider initially to comply with some protocol the markdown view laid out, but when I saw that the markdown view literally never uses that image provider (and trust me I desperately read the source code of that markdown view) I instantly dropped it and made a nice image _view_

I definitely think we should all unify and use the same image _view_ throughout the application. I definitely don't think this should be a provider.

Providers feel very anti-pattern in swift, they are less reactive, and anything they can do can be produced better and cleaner with simple views and the environment

So I completely removed that silly image provider

File: Mlem/Views/Shared/Cached Image With Nsfw Filter.swift [hunk 4]
```diff
@@ -0,0 +1,77 @@
+//
+//  MarkdownImageProvider.swift
+//  Mlem
+//
+//  Created by tht7 on 26/06/2023.
+//
+
+import Foundation
+import SwiftUI
+import MarkdownUI
```
Comment (JakeShirley) [inline Mlem/Views/Shared/Cached Image With Nsfw Filter.swift:10]: Is this used here?

**Review by JakeShirley — APPROVED** [review]

File: Mlem/Views/Shared/Cached Image With Nsfw Filter.swift (hunk 4 above)
Comment (EricBAndrews) [reply Mlem/Views/Shared/Cached Image With Nsfw Filter.swift:10]: Resolving for the sake of getting it merged, I'll check and fix if needed in upcoming post layout PR

Comment (tht7) [reply Mlem/Views/Shared/Cached Image With Nsfw Filter.swift:10]: Nooo it’s not! Dammit it’s a leftover

**Pull request merged**